UIforETW/PowerStatus.cpp

```diff
@@ -108,125 +108,179 @@ void CPowerStatusMonitor::SampleBatteryStat()
-							// With the tag, you can query the battery info.
-							BATTERY_INFORMATION bi = {};
-							bqi.InformationLevel = BatteryInformation;
-
-							if (DeviceIoControl(hBattery,
-												IOCTL_BATTERY_QUERY_INFORMATION,
-												&bqi,
-												sizeof(bqi),
-												&bi,
-												sizeof(bi),
+		// Found a battery. Query it.
+		HANDLE hBattery = CreateFile(pdidd->DevicePath,
+						GENERIC_READ | GENERIC_WRITE,
+						FILE_SHARE_READ | FILE_SHARE_WRITE,
+						NULL,
+						OPEN_EXISTING,
+						FILE_ATTRIBUTE_NORMAL,
+						NULL);
+
+		if (hBattery == INVALID_HANDLE_VALUE)
+		{
+			FreeLocalAlloc(pdidd);
+			continue;
+		}
+
+		BATTERY_QUERY_INFORMATION bqi = {};
+
+		DWORD dwWait = 0;
+		DWORD dwOut;
+
+		const BOOL tagQueried = DeviceIoControl(hBattery,
+												IOCTL_BATTERY_QUERY_TAG,
+												&dwWait,
+												sizeof(dwWait),
+												&bqi.BatteryTag,
+												sizeof(bqi.BatteryTag),
 												&dwOut,
-												NULL))
-							{
-								// Only non-UPS system batteries count
-								if (bi.Capabilities & BATTERY_SYSTEM_BATTERY)
-								{
-									// Query the battery status.
-									BATTERY_WAIT_STATUS bws = {};
-									bws.BatteryTag = bqi.BatteryTag;
-
-									BATTERY_STATUS bs;
-									if (DeviceIoControl(hBattery,
-														IOCTL_BATTERY_QUERY_STATUS,
-														&bws,
-														sizeof(bws),
-														&bs,
-														sizeof(bs),
-														&dwOut,
-														NULL))
-									{
-										char powerState[100];
-										powerState[0] = 0;
-										if (bs.PowerState & BATTERY_CHARGING)
-											strcat_s(powerState, "Charging");
-										if (bs.PowerState & BATTERY_DISCHARGING)
-											strcat_s(powerState, "Discharging");
-										if (bs.PowerState & BATTERY_POWER_ON_LINE)
-										{
-											if (powerState[0])
-												strcat_s(powerState, ", on AC power");
-											else
-												strcat_s(powerState, "On AC power");
-										}
-
-										float batPercentage = bs.Capacity * 100.f / bi.FullChargedCapacity;
-
-										char rate[100];
-										if (bs.Rate == BATTERY_UNKNOWN_RATE)
-											sprintf_s(rate, "Unknown rate");
-										else if (bi.Capabilities & BATTERY_CAPACITY_RELATIVE)
-											sprintf_s(rate, "%ld (unknown units)", bs.Rate);
-										else
-											sprintf_s(rate, "%1.3f watts", bs.Rate / 1000.0f);
-										ETWMarkBatteryStatus(powerState, batPercentage, rate);
-										//outputPrintf(L"%S, %1.3f%%, %S\n", powerState, batPercentage, rate);
-									}
-								}
-							}
-						}
-						CloseHandle(hBattery);
-					}
-				}
-				LocalFree(pdidd);
-			}
+												NULL);
+
+		if (!(tagQueried && bqi.BatteryTag))
+		{
+			FreeLocalAlloc(pdidd);
+			CloseValidHandle(hBattery);
+			continue;
+		}
+
+		// With the tag, you can query the battery info.
+		BATTERY_INFORMATION bi = {};
+		bqi.InformationLevel = BatteryInformation;
+		
+		const BOOL infoQueried = DeviceIoControl(hBattery,

```

Comment:
Every variable in scope increases complexity. Sometimes naming a result is necessary, or sometimes it increases understanding, but in most of these cases I don't think that is so. I think just "if (!DeviceIoControl(...)) continue;" would be better.